### ISSUE TITLE
[Improve Concept Docs]: Class Inheritance

### ISSUE THREAD

**Issue by BethanyG** [exercism/python#3095]
This issue describes desired additions and improvements to the `class inheritance` **concept docs.**.
The related **concept exercise** issue and specs can be found [here](https://github.com/exercism/python/issues/3096),

<br>

If you have not yet contributed to concept documents, this issue will require some upfront reading to give you the needed background knowledge.  Additionally, we recommend reading the existing `about.md` docs [here](https://github.com/exercism/python/blob/main/concepts/class-inheritance/about.md), and an example of completed concept docs [here](https://github.com/exercism/python/tree/main/concepts/classes).

<br>

## ✅   Getting started

**`Please please read the docs before starting.`** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. 

**General Contributing Docs:**

- [Contributing to Exercism](https://exercism.org/docs/building) | [Exercism and GitHub](https://exercism.org/docs/building/github) | - [Contributor Pull Request Guide](https://exercism.org/docs/building/github/contributors-pull-request-guide)
- [What are those Weird Task Tags about?](https://exercism.org/docs/building/product/tasks)
- [Exercism Formatting and Style Guide](https://exercism.org/docs/building/markdown/style-guide)
- [Exercism Markdown Specification](https://exercism.org/docs/building/markdown/markdown)
- [Reputation](https://exercism.org/docs/using/product/reputation)  

**Documents on Language Tracks and Concepts:**

- [Building Language Tracks: An Overview](https://exercism.org/docs/building/tracks) 
- [What are Concepts?](https://exercism.org/docs/building/tracks/concepts)
- [Concept Specifications](https://exercism.org/docs/building/tracks/concepts) 

<br>

## 🙏🏽   Desired Improvements and Changes

- [ ]  Create/complete the missing `introduction.md`, `links.json`, and `.meta/config.json` files for this concept.
- [ ] Where code examples show the output of running a program, they should be formatted as REPL examples (_see [this concept](https://github.com/exercism/python/blob/main/concepts/lists/about.md)  for some examples of REPL formatting_)
- [ ] Add more explanation or examples (_or argument the existing one_) of  inheriting and then extending attributes and methods from a single parent class.  [This article](https://www.pythonmorsels.com/inheriting-one-class-another/) has some good examples.
- [ ] Re-organize document to put the explanation and various uses of `super()` into a section following the **single inheritance** section.
  - [ ] Clarify/Add a bit more on what `super()` does, and why you might use it.  [This article](https://realpython.com/python-super/#an-overview-of-pythons-super-function) has some good information.  [This article](https://rhettinger.wordpress.com/2011/05/26/super-considered-super/) and [this article](https://amyboyle.ninja/Python-Inheritance) have some good info, too.
  - [ ]  Add example that illustrates the use of `super().__init__()` in making a derived class.
  - [ ]  Add example that shows the use of `super()` in  a derived class to call a parent method.
- [ ]  Provide more explanation for multiple inheritance, and some background on its controversy.  (_[This article](https://www.thedigitalcatonline.com/blog/2020/03/27/mixin-classes-in-python/#multiple-inheritance-blessing-and-curse) has some good sections explaining multiple inheritance and its issues_). 
  - [ ]   Provide a practical example of multiple inheritance with extended attributes and methods in the child class.  [This article](https://www.pythontutorial.net/python-oop/python-multiple-inheritance/) might have some good examples, as well as [this one](https://baransel.dev/post/python-multiple-inheritance/)
  - [ ]   Provide a practical example of multiple inheritance **_problems_** in action, as a set up to the diamond inheritance problem. [This article](https://medium.com/@varunsingh-02/master-python-inheritance-in-2022-with-these-tricks-aa0becfbbd54) might have some examples that help.
  - [ ]   Clarify the diamond problem explanation, and talk a little about [Python's MRO](https://levelup.gitconnected.com/method-resolution-order-in-python-5afbaecc25e0) and mention its [C3 algorithm](https://en.wikipedia.org/wiki/C3_linearization) for multiple inheritance resolution.  This is an old but good [blogpost about MRO](http://python-history.blogspot.com/2010/06/method-resolution-order.html).
- [ ]  Add how `super()` can help with the diamond problem, and how `super()` affects the MRO in multiple inheritance.  [This section of the Real Python article on `super()` can help](https://realpython.com/python-super/#super-in-multiple-inheritance), as can [this article](https://rhettinger.wordpress.com/2011/05/26/super-considered-super/).
- [ ]  Clarify what a `mixin` **is** (_a class created solely to add specific and isolated functionality to one or more base classes_) and what it can do generally.  [This SO post](https://stackoverflow.com/questions/533631/what-is-a-mixin-and-why-is-it-useful),  [This article](https://dev.to/bikramjeetsingh/write-composable-reusable-python-classes-using-mixins-6lj) and [this article](https://andrewbrookins.com/technology/mixins-in-python-and-ruby-compared/#what-is-a-mixin) can help.
  - you want to provide a lot of optional features for a class
  - you want to provide a specific set of features in a lot of different classes
- [ ]   Clarify that `Mixins` are often used to combat "gotchas" that come with multiple inheritance.  [this article](https://coderbook.com/@marcus/deep-dive-into-python-mixins-and-multiple-inheritance/) has some good information for that.
- [ ]  Talk about the Python convention of ending  a `mixin` class name with "Mixin" 
- [ ]   Show a practical example of a `mixin` class and its use in one or more child classes.

<br>

_________

# 🌟   Original Specifications

Below are the original specifications from the Concept issue.  Please use these as a guide in addition to the change list above.

## 🎯   Goal

This concept is meant to teach an understanding/use of `class-inheritance` in Python.

<br>

## 💡   Learning objectives

- more fully understand the uses (_and possible abuses_) of `classes`  in Python
- understanding and use of `single inheritance` in Python
- understanding and use of `multiple inheritance` in Python
  - the diamond inheritance problem
  - Pythons MRO (_method resolution order_)
  - pros and cos of `mixins` () to isolate `multiple inheritance` issues
- when and when not to use `super()` to call a parent `class` constructor or method

<br>

## 🚫  Out of scope

<details>
<summary><b>Concepts and Subjects that are Out of Scope</b></summary>
<br>

- `class-composition` as an explicit topic (_it is ok to use limited composition as needed in example code_)
- `dataclasses`
- `decorators` outside of `@property`, `@staticmethod`, and `@classmethod` (_the student should already be exposed to these in the prerequisite exercise_)
- `generators`
- `coroutines`
- `descriptors` (_these will get their own exercise_)
- using a `class` as a decorator
- performance considerations

</details>

<br>

## 🤔  Concepts

<details>
<summary><b>Concepts and Related Concepts this Concept Covers</b></summary>
<br>

- `classes`
- `inheritance` in `OOP`
- `single inheritance`
- `multiple inheritance`
- `minxins`
- `super()`

</details>

<br>

## ↩️   Prerequisites

These are the concepts/concept exercises the student should be familiar with before taking on/learning this concept.  

<details>
<summary>Prereqs</summary>
<br>

- `basics`
- `booleans`
- `classes`
- `class-customization`
- `comparisons`
- `decorators`
- `dicts`
- `functions`
- `higher-order-functions`
- `iteration`
- `lists`
- `numbers`
- `sequences`
- `sets`
- `strings`
- `tuples`

</details>

<br>  

## 📚  Resources to Refer to (_in addition to the ones linked in the task list_)

<details>
<summary><b>Resources</b></summary>
<br>

- [classes (Python tutorial)](https://docs.python.org/3/tutorial/classes.html)
- [Real Python: Object-Oriented Programming in Python 3](https://realpython.com/python3-object-oriented-programming/)
- [Real Python: Inheritance and Composition: A Python OOP Guide](https://realpython.com/inheritance-composition-python/)
- [Towards DataScience: Understand Inheritance in Python](https://towardsdatascience.com/understand-inheritance-in-python-74f8e3025f3c) -- (_this is not a good resource for hints or links, since it requires an account signup_)
  - [Liskov substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle)
- [Multiple Inheritance in Python](https://www.python-course.eu/python3_multiple_inheritance.php)
- [Multiple Inheritance is Hard (Ned Batchelder)](https://nedbatchelder.com/blog/201210/multiple_inheritance_is_hard.html)
- [Python.org: MRO in Python 2.3+](https://www.python.org/download/releases/2.3/mro/)
- [Dig into Python super() and MRO](https://gaopinghuang0.github.io/2018/12/29/dig-into-python-super-and-mro)
- [Multiple inheritance and mixin classes in Python](https://www.thedigitalcatonline.com/blog/2020/03/27/mixin-classes-in-python/)
- [Making Python classes more modular using mixins](https://www.residentmar.io/2019/07/07/python-mixins.html)
- [Write composable, reusable Python classes using Mixinsl)](https://dev.to/bikramjeetsingh/write-composable-reusable-python-classes-using-mixins-6lj)
- [Mixins in Python and Ruby Compared](https://andrewbrookins.com/technology/mixins-in-python-and-ruby-compared/)
- [Python Data Model - Python Docs](https://docs.python.org/3/reference/datamodel.html#the-standard-type-hierarchy)

</details>
<br>

##  📁  Files to Be Updated or Created

<details open>
<summary><b>File Detail for this Exercise</b></summary>
<table>
<td>
<br>

Please see the following for more details on these files:  [concepts](https://exercism.org/docs/building/tracks/concepts)

* ### `links.json`

  For more information, see [concept links file](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-linksjson)

  -  The same resources listed in this issue can be used as a starting point for the [ `concepts/links.json`](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-linksjson)  file, if it doesn't already exist.
  -  If there are particularly good/interesting information sources for this concept that extend or supplement the concept exercise material & the resources already listed -- please add them to the `links.json` document.

* ### Concept `about.md`

  For more information, see [Concept `about.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-aboutmd)

  - This file provides information about this concept for a student who has completed the corresponding concept exercise.  It is intended as a reference for continued learning.  

* ### Concept `introduction.md`

  For more information, see [Concept `introduction.md`](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-aboutmd)

  * This can also be a summary/paraphrase of the  `about.md` document listed above, and will provide a brief introduction of the concept for a student who has **not yet** completed the associated concept or practice exercises.   It should contain a good summation of the concept, but not go into lots of detail.

* ### Concept `.meta/config.json` Entries
 
  For more information, see [Concept `.meta/config.json`](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-metaconfigjson)

  * This file is likely already stubbed out.  Remember to add a concept **blurb** of less than 350 characters.  Please also add your GitHub username to the "authors" array, and any contributor GitHub usernames to the "contributors" array.

<br>
</td>
</table>
</details>

## 🎶  Implementation Notes

- Example code  should **only use syntax & concepts introduced within these docs or one of the prerequisite concept exercises or documents.** Where possible, please use REPL formatting, unless you are demonstrating pseudo code or a long code block.
  Please **do not use** syntax not previously covered in prerequisite topics or exercises.  Please also follow [PEP8](https://www.python.org/dev/peps/pep-0008/) guidelines.
- Our markdown and JSON files are checked against [prettier](https://prettier.io/) .  We recommend [setting prettier up locally](https://prettier.io/docs/en/install.html) and running it prior to submitting your PR  to avoid any CI errors.

<br>

##  🆘  Next Steps & Getting Help

1.  **`If you'd like to work on this issue, comment saying "I'd like to work on this"`** (_there is no real need to wait for a response, just go ahead, we'll assign you and put a `[claimed]` label on the issue_).
2. If you have any questions while implementing, please post the questions as comments in here, or contact one of the maintainers on our Slack channel.
